src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,133 @@
+            /// can still support when <see cref="ApplicableActions"/> is empty. It's desirable
+            /// that consumers can call a rename API to produce a <see cref="RenameDocumentActionSet"/> and
+            /// immediately call <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> without
+            /// having to inspect the returned <see cref="ApplicableActions"/>.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (solution is null)
+                {
+                    throw new ArgumentNullException(nameof(solution));
+                }
+
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(string.Format(WorkspacesResources.Cannot_apply_action_that_is_not_in_0, nameof(ApplicableActions)));
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
+
+                // Apply each action individually. Order should not matter
+                foreach (var action in actions)
+                {
+                    document = solution.GetRequiredDocument(documentId);
+                    solution = await action.GetModifiedSolutionAsync(document, _optionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Attempts to find the document in the solution. Tries by documentId first, but 
+            /// that's not always reliable between analysis and application of the rename actions
+            /// </summary>
+            private Document GetDocument(Solution solution)
+            {
+                // DocumentId is the best bet for finding a document,
+                // but it's possible the document was renamed or moved
+                // before actions were applied.
+                if (solution.ContainsDocument(_documentId))
```

Comment:
can't we rename while keeping the doc-id?  @jasonmalinowski

Reply:
Ideally yes, but we don't currently have a way. Right now it comes across as a remove/add

Reply:
i guess i'm really confused by this.  what is the case where hte documentId changes... but the doc name and doc folders are the same?  i.e. the 'else' branch seems really suspect to hit.  can we actually doc exactly when this would happen and why we need to support it?